ui/src/frontend/scroll_helper.ts

```diff
@@ -121,64 +121,105 @@ export function verticalScrollToTrack(
+  }
+}
+
+// Get the rendered track groups in the given hierarchical chain,
+// in order from topmost to deepest, as many as are rendered by
+// their own container having been expanded in the UI.
+// These are represented in the resulting array by |Element|.
+// All further nested track groups that are not rendered are
+// represented in the resulting array by their UUIDs.
+// down, represented either as the rendered HTML element or, if not
+// rendered because an ancestor group is collapsed, its UUID.
+function getTrackGroupsTopDown(groupIds: string[]): (Element|string)[] {
+  const result: (Element|string)[] = [];
+
+  for (const trackGroupId of groupIds) {
     const group = document.querySelector('#track_' + trackGroupId);
 
     if (group) {
-      trackGroupsTopDown.push(group);
+      result.push(group);
     } else {
       // Some containing group is collapsed, so this one is not rendered
-      trackGroupsTopDown.push(trackGroupId);
+      result.push(trackGroupId);
     }
   }
 
-  // The requested track is inside a closed track group. Either open the track
-  // group and scroll to the track or just scroll to deepest nested track group
-  // that has been rendered.
-  if (openGroup) {
-    // After the track exists in the dom, it will be scrolled to.
-    globals.frontendLocalState.scrollToTrackId = trackId;
-    for (let i = 1; i < trackGroupsTopDown.length; i++) {
-      const trackGroup = trackGroupsTopDown[i];
-      if (typeof trackGroup === 'string') {
-        // Expand its parent
-        if (typeof trackGroupsTopDown[i - 1] !== 'string') {
-          const trackGroupId = containingIds[i - 1];
-          globals.dispatch(Actions.toggleTrackGroupCollapsed({trackGroupId}));
-        }
-
-        // And mark all the rest for expansion when they are created
-        globals.frontendLocalState.expandTrackGroupIds.add(trackGroup);
-      } else if (i === (trackGroupsTopDown.length - 1)) {
-        // Need to expand the bottommost group to create the track to reveal
-        const trackGroupId = containingIds[i];
+  return result;
+}
+
+// Expand whatever groups in the |trackGroupsTopDown| that are
+// collapsed as necessary to reveal and scroll to the track
+// identified by the given |trackId|. Each element of the
+// |trackGroupTopDown| is identified by the corresponding UUID
+// in the |containingGroupIds| array. As there is potentially
+// some tail of the former that are UUIDs representing unrendered
+// groups (because their container is collapsed), there may be
+// overlap in these arrays.
+function expandGroupsToRevealTrack(trackId: string,
+    containingGroupIds: string[],
+    trackGroupsTopDown: (Element|string)[]): void {
+  // After the track exists in the DOM, it will be scrolled to.
+  globals.frontendLocalState.scrollToTrackId = trackId;
+
+  trackGroupsTopDown.forEach(
+      (trackGroup: Element|string, i: number) => {
+    // It should not happen that the topmost group is not rendered
```

Comment:
And of course, you could do what I recommend without asserts, but it wouldn't be succinct, and it wouldn't be in a way that could support disabling such checks in a production environment, if that's what we'd like to do.

Reply:
Commit eac7b387e22849ef5435837ec9315ef20691d608 adds Perfetto-style assertions. If we want to make them optional at run-time, that could be done then in the assertion API.

Reply:
Thanks, @cdamus . BTW, the topic at hand was assertion elsewhere. I was aware that perfetto itself uses asserts. But in discussing using asserts in Sokatoa in general, I got the sense that maybe the value-add wasn't clear. I was just trying to point out to Martin an example of where I think they would help. 

Thanks for adding these to your change!

Reply:
I see what you mean with the `assertion` support and since it is a very simple shorthand I have absolutely nothing against it. I'm sure it will help us out in a lot of places.